presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -161,7 +161,8 @@ public void finish()
     @Override
     public boolean isFinished()
     {
-        boolean finished = this.finished && probe == null && pageBuilder.isEmpty() && outputPage == null;
+        boolean finished = (this.finished && probe == null && pageBuilder.isEmpty() && outputPage == null) ||
+                isLookupSourceEmpty();
```

Comment:
Instead of inspecting repeatedly whether lookup source is empty, you could leverage the fact that this information doesn't change, ever.

So, instead of this change (and instead of introducing the `isLookupSourceEmpty` method), you can inspect `lookupSource` only once, in `tryFetchLookupSourceProvider`, and set `finishing` flag when the LS is empty.

Reply:
Good suggestion! isLookupSourceEmpty() is somewhat convoluted and I share the sentiment. Let me work on the change so it just one call to tryFetchLookupSourceProvider.